src/Disciplina/Crypto/Signing/Class.hs

```diff
@@ -14,15 +16,27 @@ import Universum
 
 import Data.ByteArray (ByteArrayAccess)
 
+-- | Class of signature schemes with defined format of keys and
+-- signatures.
+class SignatureScheme ss where
```

Comment:
Are you sure we won't have any cool ideas about using some, like, hash-based signatures in the future?
Also, I tried to make these modules as generic as possible, because I plan to move them to `lootbox` when they look alright (and to reuse them elsewhere). Sorry that I didn't make this intention clear.

Reply:
@volhovm No, this isn't the only reason we want these wrappers.
1) Phantom type parameters for type safety and shit.
2) Kind-of-smart approach to serialization. In `cardano-sl`, every value which is going to be hashed or signed is serialized first, which requires memory allocation and copying - even if hashed/signed value is _already_ a byte array. For example, it happens every time when we sign a hash of transaction - hash gets copied and serialized again for no reason.

You may argue that "this optimization will turn out to be very minor in real life" or "shouldn't we just serialize all values manually before signing/hashing?". For checking the first statement, we should probably find some benchmark results for `cardano-sl` and look at how much time and memory is consumed by serialization. The answer to a second question is "no, we shouldn't, this will make everybody's life slightly more miserable".

Reply:
> Phantom type parameters for type safety and shit.
This is not an argument, you could use phantom types with specific `newtype` wrappers without using a generalised class.

> Kind-of-smart approach to serialization.
That's why one should use `Hashable` constraint (our own one, I mean, not the one that is used in `HashMap`) instead of `Binary`, it solves the issue. First, you implement a default `hash` method using `Binary`, which applies a hash after encoding. Second, for `BS`-compatible types you implement your own method instead of the default, which uses `hash` directly. Voila.

>The first "you may argue" statement
It is hard to be sure without checking the real data, yes. Anyway, I can't really think of any real usecase of "taking a hash of the bytestring". We usually hash block body components (which are complex datatypes) to generate a block proof, we also hash addresses, but they are complex too. 

>The second "you may argue" statement
I wouldn't say that, it's way too straight-forward; see (2) in this comment.

Still not conceived.

Reply:
>First, you implement a default hash method using Binary, which applies a hash after encoding. Second, for BS-compatible types you implement your own method instead of the default, which uses hash directly. Voila.

This is _exactly_ what I'm doing, nothing more. Also, this suggestion contradicts the phrase "without using a generalized class" from your previous statement, because this `Hashable` is exactly this generalized class.

Reply:
`Hashable` class only has one method, which turns your type into something that can be hashed. `SignatureScheme` has two methods and three type family declarations and is designed to be used with several signature schemes, which is exactly what I am opposed to.

We can merge it though; if you don't see the difference, the entire conversation doesn't make sense.